apps/opencs/view/render/terraintexturemode.cpp

```diff
@@ -0,0 +1,328 @@
+{
+    mBrushTextureLabel = "Brush: " + mBrushTexture;
+    selectedBrush = new QLabel(QString::fromUtf8(mBrushTextureLabel.c_str()), this);
+
+    QVBoxLayout *layoutMain = new QVBoxLayout;
+    layoutMain->setSpacing(0);
+
+    QHBoxLayout *layoutHorizontal = new QHBoxLayout;
+    layoutHorizontal->setContentsMargins (QMargins (0, 0, 0, 0));
+    layoutHorizontal->setSpacing(0);
+
+    configureButtonInitialSettings(buttonPoint);
+    configureButtonInitialSettings(buttonSquare);
+    configureButtonInitialSettings(buttonCircle);
+    configureButtonInitialSettings(buttonCustom);
+
+    QButtonGroup* brushButtonGroup = new QButtonGroup(this);
+    brushButtonGroup->addButton(buttonPoint);
+    brushButtonGroup->addButton(buttonSquare);
+    brushButtonGroup->addButton(buttonCircle);
+    brushButtonGroup->addButton(buttonCustom);
+
+    brushButtonGroup->setExclusive(true);
+
+    layoutHorizontal->addWidget(buttonPoint);
+    layoutHorizontal->addWidget(buttonSquare);
+    layoutHorizontal->addWidget(buttonCircle);
+    layoutHorizontal->addWidget(buttonCustom);
+
+    horizontalGroupBox = new QGroupBox(tr(""));
+    horizontalGroupBox->setLayout(layoutHorizontal);
+
+    BrushSizeControls* sizeSliders = new BrushSizeControls(tr(""), this);
+
+    layoutMain->addWidget(horizontalGroupBox);
+    layoutMain->addWidget(sizeSliders);
+    layoutMain->addWidget(selectedBrush);
+
+    setLayout(layoutMain);
+
+    connect(buttonPoint, SIGNAL(clicked()), this, SLOT(setBrushShape()));
+    connect(buttonSquare, SIGNAL(clicked()), this, SLOT(setBrushShape()));
+    connect(buttonCircle, SIGNAL(clicked()), this, SLOT(setBrushShape()));
+    connect(buttonCustom, SIGNAL(clicked()), this, SLOT(setBrushShape()));
+
+    connect(sizeSliders->brushSizeSlider, SIGNAL(valueChanged(int)), parent, SLOT(setBrushSize(int)));
+
+    connect(parent, SIGNAL(passBrushTexture(std::string)), this, SLOT(getBrushTexture(std::string)));
+}
+
+void CSVRender::TextureBrushWindow::configureButtonInitialSettings(TextureBrushButton *button)
+{
+  button->setSizePolicy (QSizePolicy (QSizePolicy::Fixed, QSizePolicy::Fixed));
+  button->setContentsMargins (QMargins (0, 0, 0, 0));
+  button->setIconSize (QSize (48-6, 48-6));
+  button->setFixedSize (48, 48);
+  button->setAcceptDrops(true);
+  button->setCheckable(true);
+}
+
+void CSVRender::TextureBrushWindow::getBrushTexture(std::string brushTexture)
+{
+    mBrushTexture = brushTexture;
+    mBrushTextureLabel = "Brush:" + mBrushTexture;
+    selectedBrush->setText(QString::fromUtf8(mBrushTextureLabel.c_str()));
+}
+
+void CSVRender::TextureBrushWindow::setBrushSize(int brushSize)
+{
+    mBrushSize = brushSize;
+    emit passBrushSize(mBrushSize);
+}
+
+void CSVRender::TextureBrushWindow::setBrushShape()
+{
+    if(buttonPoint->isChecked()) mBrushShape = 0;
+    if(buttonSquare->isChecked()) mBrushShape = 1;
+    if(buttonCircle->isChecked()) mBrushShape = 2;
+    if(buttonCustom->isChecked()) mBrushShape = 3;
```

Comment:
This whole construct looks suboptimal. But I need to take a closer look before I can give any advice on it.